### PR TITLE
Moves Graph Atlas to data file.

### PR DESCRIPTION
**Pull request by jfinkels**
This commit moves the edge lists of the graphs in the graph atlas from a
Python list defined in `networkx/generators/atlas.py` to a text file in
the same directory. This means that accessing the list of atlas graphs
will cause a file read.

This also adds a new function, `graph_atlas(i)`, that outputs only the
`i`th graph from the atlas.

This should help fix issue #1671.

### CODE REVIEW

File: networkx/generators/tests/test_atlas.py [hunk 1]
```diff
@@ -1,62 +1,75 @@
...
-        # note that 111223 < 112222
-        previous=self.GAG[0]
-        for g in self.GAG:
-            if len(g)==0: 
-                continue
-            if len(g)==len(previous) & g.size()==previous.size():
-                deg_seq=sorted(g.degree().values())
-                previous_deg_seq=sorted(previous.degree().values())
-                assert_true(previous_deg_seq < deg_seq)
-                previous=g.copy()
```
Comment (jfinkels) [inline networkx/generators/tests/test_atlas.py]: In the old test, this line should not have been indented here, it should have been at the same level as the `if ...` above. This caused the test to pass trivially, as it wasn't checking any of the assertions. Dedenting this line reveals the same test failure as my new test below.

**Comment by jfinkels** [comment]
The order of the graphs in `graph_atlas_g()` is not as documented: there is a pair of graphs, graphs 55 and 56, that have the same number of nodes and same number of edges, but have reversed degree sequences (according to lexicographic order), i.e. the first graph has a larger degree sequence than the second graph. See my comment on the diff for why the original unit test missed this.

**Comment by jfinkels** [comment]
We could also distribute the compressed file: a simple `gzip atlas.dat` compresses a 71 kilobyte file into an 8 kilobyte file.

**Comment by hagberg** [comment]
I'm not too worried about the 63K, fine to compress.  The ordering of the graphs should be fixed though.

**Comment by jfinkels** [comment]
I'm sorry, I didn't understand exactly what you were suggesting in the above comment. Are you suggesting that I  reorder the graphs in the data file so that they are in the order described by the documentation? Or are you suggesting that I find the book *An Atlas of Graphs*, assuming the current order matches that of the graphs in the book, and change the documentation?

**Comment by hagberg** [comment]
I suppose it makes most sense to be like in the book.  Here are graph 55 and 56 https://books.google.com/books?id=QOPuAAAAMAAJ&focus=searchwithinvolume&q=g55

**Comment by jfinkels** [comment]
Cool! Do you happen to know if the book does actually state that the graphs are ordered in the way given in our documentation? In other words, is the book wrong about how its graphs are listed? (If you don't actually have a copy, I can find the book in the library.)

It would be nice if the documentation for NetworkX was able to describe precisely how the graphs are ordered. Is there another way of interpreting "ordered by increasing degree sequence"? Maybe I'm just not understanding it correctly.

**Comment by hagberg** [comment]
I might be able to find a book.  Maybe @pieterswart could tell us about the graph ordering - he wrote the atlas module.

**Comment by chebee7i** [comment]
Oh Read! Nice. I haven't read that book, but Read has some other work about orderly generation of graphs which I'm sure this book is based on to some degree (pun intended). The generation order is a bit arbitrary in some cases if I recall correctly but well-defined and designed to avoid having to perform an isomorphism check of new candidate graphs against all previously generated graphs.

If you can find this paper: "Every One A Winner; or How to avoid isomorphism search when cataloguing combinatorial configurations", it might have the info you need. I have a copy on my machine at home, so let me know if access is difficult.

**Comment by pieterswart** [comment]
Going back many years, here is the description I have:

Re the ordering in Graph Atlas:

the graphs are listed:

1) in increasing order of number of nodes;
2) for a fixed number of nodes, in increasing order of the number of edges;
3) for fixed numbers of nodes and edges, in increasing order of the degree
sequence, for example 111223 < 112222;
4) for fixed degree sequence, in increasing number of automorphisms.
5) In NetworkX, indexing is set up so that for GAG=graph_atlas_g(),
   then G123=GAG[123] and G[0]=empty_graph(0)

Pieter Swart

**Comment by jfinkels** [comment]
I looked at a copy of the book in a library, and what Pieter wrote does match the text. Examining the rest of the list of graphs, they do generally seem to match that description (i.e. sorted ascending by number of vertices, number of edges, degree sequence, and number of automorphisms, in that order). However, the graphs numbered 55 and 56 fail to meet that sort order. I believe this to be an error in the book.

I propose changing the documentation to note this exception to the sorting rules (and others, should they exist) and change the unit tests to exclude this particular pair of graphs.

**Comment by hagberg** [comment]
@jfinkels  I agree with your proposal

**Comment by jfinkels** [comment]
Great, I'll work on updating this pull request.

**Comment by jfinkels** [comment]
I've updated the documentation in `graph_atlas_g` to note the exceptions. There are three exceptions to the lexicographic sort order rule.

**Pull request merged**